### PR TITLE
Make properties which are part of PrimaryKey as required

### LINKED ISSUE #1978: Conventions: Make key properties required
Problem seems to be that upon rename we create nullable column which then can't be made a Primary key.

Configuration before:

```

    public class GearsOfWarContext : DbContext
    {
        public GearsOfWarContext()
        {
        }

        public DbSet<City> Cities { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(new SqlConnectionStringBuilder { DataSource = ".", InitialCatalog = "Beta4Bash_Migrations", MultipleActiveResultSets = true, IntegratedSecurity = true }.ToString());
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<City>().Key(e => e.Name).ForRelational().Name("Foobar");
            modelBuilder.Entity<City>().Property(e => e.LocationChange);
            modelBuilder.Entity<City>().Property(e => e.LocationChange).MaxLength(40);

            modelBuilder.Entity<City>().Property<DateTime>("Shadow");
        }
    }

    public class City
    {
        public string Name { get; set; }
        public int LocationChange { get; set; }
    }
```

Configuration after:

```

    public class GearsOfWarContext : DbContext
    {
        public GearsOfWarContext()
        {
        }

        public DbSet<City> Cities { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(new SqlConnectionStringBuilder { DataSource = ".", InitialCatalog = "Beta4Bash_Migrations", MultipleActiveResultSets = true, IntegratedSecurity = true }.ToString());
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<City>().Key(e => e.NameId).ForRelational().Name("Foobar");
            modelBuilder.Entity<City>().Property(e => e.LocationChange);
            modelBuilder.Entity<City>().Property(e => e.LocationChange).MaxLength(40);

            modelBuilder.Entity<City>().Property<DateTime>("Shadow");
        }
    }

    public class City
    {
        public string NameId { get; set; }
        public int LocationChange { get; set; }
    }
```

Migration:

```
        public override void Up(MigrationBuilder migration)
        {
            migration.DropPrimaryKey(name: "Foobar", table: "City");
            migration.DropColumn(name: "Name", table: "City");
            migration.AddColumn(
                name: "NameId",
                table: "City",
                type: "nvarchar(450)",
                nullable: true);
            migration.AddPrimaryKey(
                name: "Foobar",
                table: "City",
                column: "NameId");
        }
        
        public override void Down(MigrationBuilder migration)
        {
            migration.DropPrimaryKey(name: "Foobar", table: "City");
            migration.DropColumn(name: "NameId", table: "City");
            migration.AddColumn(
                name: "Name",
                table: "City",
                type: "nvarchar(450)",
                nullable: true);
            migration.AddPrimaryKey(
                name: "Foobar",
                table: "City",
                column: "Name");
        }
```

Exception:

System.Data.SqlClient.SqlException (0x80131904): Cannot define PRIMARY KEY constraint on nullable column in table 'City'.
At line:1 char:1
+ dnx . ef migration apply
+ ~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (System.Data.Sql...n table 'City'.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
Could not create constraint. See previous errors.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean 
breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean async, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, String methodName, Boolean sendToPipe, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Microsoft.Data.Entity.Relational.SqlStatementExecutor.<>c__DisplayClass7_0.<ExecuteNonQuery>b__0()
   at Microsoft.Data.Entity.Relational.SqlStatementExecutor.Execute(IRelationalConnection connection, Func`1 action)
   at Microsoft.Data.Entity.Relational.SqlStatementExecutor.ExecuteNonQuery(IRelationalConnection connection, DbTransaction transaction, IEnumerable`1 sqlBatches)
   at Microsoft.Data.Entity.Relational.Migrations.Migrator.Execute(IEnumerable`1 sqlBatches, Boolean ensureDatabase)
   at Microsoft.Data.Entity.Relational.Migrations.Migrator.ApplyMigrations(String targetMigration)
   at Microsoft.Data.Entity.Commands.MigrationTool.ApplyMigration(String migrationName, String contextTypeName)
   at Microsoft.Data.Entity.Commands.Program.<>c__DisplayClass10_0.<ApplyMigration>b__0()
   at Microsoft.Data.Entity.Commands.Program.ExecuteInDirectory(String startupProject, Func`1 invoke)
   at Microsoft.Data.Entity.Commands.Program.ApplyMigration(String migration, String context, String startupProject)
   at Microsoft.Data.Entity.Commands.Program.<>c__DisplayClass7_2.<Main>b__8()
   at Microsoft.Framework.Runtime.Common.CommandLine.CommandLineApplication.Execute(String[] args)
   at Microsoft.Data.Entity.Commands.Program.Main(String[] args)
ClientConnectionId:73c919fd-63dc-4c08-8b6d-ed1f32af40dc

### PR DESCRIPTION
**Pull request by smitpatel**
fixes #1978 
Property which are part of PrimaryKey will be required by default

### CODE REVIEW

**Comment by lajones** [comment]
It depends on the usage of course, but it can be useful to allow models in bad states. This allows you to represent models which are in bad or intermediate states in the process of making them good. For instance I have already come across a situation in ReverseEngineering where I want to represent an entity without a primary key - because e.g. the primary key was of a SQL Server type for which we don't have a mapping but I still want to output the file based on the entity but containing an error message explaining about the property instead. And, looking to the future, when representing intermediate states during updating the model based on an updated DB it can be useful to allow states that would normally be "illegal". But so long as there is some way of doing that then maybe it's fine to do as @ajcvickers suggests here.

**Comment by ajcvickers** [comment]
Discussed with @divega and decided we should do it in the Property. Look at how the state for ReadOnlyAfterSave is handled and do the same for Required.

File: src/EntityFramework.Core/Metadata/Property.cs [hunk 1]
```diff
@@ -42,17 +42,23 @@ public Property([NotNull] string name, [NotNull] Type clrType, [NotNull] EntityT
...
-                    && value.Value
-                    && !ClrType.IsNullableType())
+                    && value.Value)
                 {
-                    throw new InvalidOperationException(Strings.CannotBeNullable(Name, EntityType.DisplayName(), ClrType.Name));
+                    if (!ClrType.IsNullableType())
+                    {
+                        throw new InvalidOperationException(Strings.CannotBeNullable(Name, EntityType.DisplayName(), ClrType.Name));
+                    }
+                    if (EntityType.FindPrimaryKey() != null && EntityType.FindPrimaryKey().Properties.Contains(this))
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Property.cs]: Consider refactoring to make only one call to FindPrimaryKey

File: src/EntityFramework.Core/Metadata/Property.cs [hunk 2]
```diff
@@ -42,17 +42,23 @@ public Property([NotNull] string name, [NotNull] Type clrType, [NotNull] EntityT
...
+                        throw new InvalidOperationException(Strings.CannotBeNullablePK(Name, EntityType.DisplayName()));
+                    }
                 }
 
                 SetFlag(value, PropertyFlags.IsNullable);
             }
         }
 
-        protected virtual bool DefaultIsNullable => ClrType.IsNullableType();
+        protected virtual bool DefaultIsNullable => (EntityType.FindPrimaryKey() == null || !EntityType.FindPrimaryKey().Properties.Contains(this)) && ClrType.IsNullableType();
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Property.cs]: Likewise.

File: src/EntityFramework.Core/Metadata/Property.cs [hunk 3]
```diff
@@ -168,7 +174,7 @@ public virtual int Index
             }
         }
 
-        public virtual object SentinelValue { get; [param: CanBeNull] set; }
+        public virtual object SentinelValue { get;[param: CanBeNull] set; }
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Property.cs]: Add back space before attribute. Consider using R# for formatting.

**Comment by ajcvickers** [comment]
:shipit:

**Pull request merged**